k-distribution/src/main/scripts/lib/pyk/kast.py

```diff
@@ -1,216 +1,360 @@
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {'node': 'KToken', 'token': self.token, 'sort': self.sort}
+
+    def copy(self, *, token: Optional[str] = None, sort: Optional[str] = None) -> 'KToken':
+        token = token if token is not None else self.token
+        sort = sort if sort is not None else self.sort
+        return KToken(token=token, sort=sort)
+
+
+@final
+@dataclass(frozen=True)
+class KApply(KInner):
+    label: str
+    args: Tuple[Any, ...]  # TODO strengthen to Tuple[K...]
+
+    def __init__(self, label: str, args: Iterable[Any] = ()):  # TODO strengthen
+        object.__setattr__(self, 'label', label)
+        object.__setattr__(self, 'args', tuple(args))
+
+    @property
+    def arity(self) -> int:
+        return len(self.args)
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {'node': 'KApply', 'label': self.label, 'args': [KAst._to_dict(arg) for arg in self.args], 'arity': self.arity, 'variable': False}
+
+    def copy(self, *, label: Optional[str] = None, args: Optional[Iterable[Any]] = None) -> 'KApply':  # TODO strengthen
+        label = label if label is not None else self.label
+        args = args if args is not None else self.args
+        return KApply(label=label, args=args)
 
+    def map(self, f: Callable[[Any], Any]) -> 'KApply':  # TODO strengthen
+        # TODO maybe add mapper for label as well, analogously to copy
+        return self.copy(args=(f(arg) for arg in self.args))
 
-def KAtt(atts={}):
-    return {'node': 'KAtt', 'att': atts}
 
+@final
+@dataclass(frozen=True)
+class KAs(KInner):
+    pattern: Any  # TODO strengthen
+    alias: Any    # TODO strengthen
+    att: KAtt
 
-def isKAtt(k):
-    return k['node'] == 'KAtt'
+    def __init__(self, pattern: Any, alias: Any, att=KAtt()):  # TODO strengthen
+        object.__setattr__(self, 'pattern', pattern)
+        object.__setattr__(self, 'alias', alias)
+        object.__setattr__(self, 'att', att)
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {'node': 'KAs', 'pattern': KAst._to_dict(self.pattern), 'alias': KAst._to_dict(self.alias), 'att': self.att.to_dict()}
+
+    def copy(self, *, pattern: Optional[Any] = None, alias: Optional[Any] = None, att: Optional[KAtt] = None) -> 'KAs':  # TODO strengthen
+        pattern = pattern if pattern is not None else self.pattern
+        alias = alias if alias is not None else self.alias
+        att = att if att is not None else self.att
+        return KAs(pattern=pattern, alias=alias, att=att)
+
+
+def KConstant(label):
+    return KApply(label, [])
+
+
+def isKConstant(k):
+    return type(k) is KApply and k.arity == 0
+
+
+def KSequence(items):
+    return {'node': 'KSequence', 'arity': len(items), 'items': items}
+
+
+def isKSequence(k):
+    return type(k) is dict and k['node'] == 'KSequence'
+
+
+def KRewrite(lhs, rhs):
+    return {'node': 'KRewrite', 'lhs': lhs, 'rhs': rhs}
+
+
+def isKRewrite(k):
+    return type(k) is dict and k['node'] == 'KRewrite'
 
```

Comment:
Everything to this point is `KInner`, and everything after it is `KOuter`.

It may be good to pull out a separate `kast/kinner.py`, for example.

Reply:
Let's get `KInner` fleshed out in this PR, and we can do `KOuter` in a follow-up PR.

Unless you already have work for both, and then we can do it all in 1 shot.

But just in terms of being able to merge incremental progress, so that I can make sure the submodule updates go through, may make sense to merge this with just `KInner` first.